File: pilot/src/api/adapters.rs
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::IntoResponse;
use axum::Json;
use k8s_openapi::api::core::v1::{
    Container, ContainerPort, EnvVar, Pod, PodSpec, Service, ServicePort, ServiceSpec,
};
use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
use k8s_openapi::apimachinery::pkg::util::intstr::IntOrString;
use kube::api::{Api, DeleteParams, PostParams};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;
use tracing::{info, warn};
use uuid::Uuid;

use crate::app_state::AppState;
use super::error::ApiError;

#[derive(Debug, Deserialize)]
pub struct CreateAdapterRequest {
    pub name: String,
    pub image: String,
    #[serde(default)]
    pub env: std::collections::HashMap<String, String>,
}

#[derive(Debug, Serialize)]
pub struct AdapterResponse {
    pub id: String,
    pub name: String,
    pub image: String,
    pub env: serde_json::Value,
    pub created_at: String,
}

pub async fn create(
    State(state): State<AppState>,
    Json(req): Json<CreateAdapterRequest>,
) -> Result<impl IntoResponse, ApiError> {
    if state.sqlite.get_adapter_by_name(&req.name).await?.is_some() {
        return Err(ApiError::Conflict(format!("adapter '{}' already exists", req.name)));
    }

    let id = format!("adp-{}", Uuid::now_v7());
    let env_json = serde_json::to_string(&req.env)
        .map_err(|e| ApiError::BadRequest(e.to_string()))?;

    // Step 1: Save to DB
    let adapter = state.sqlite.create_adapter(&id, &req.name, &req.image, &env_json).await?;

    // Step 2: Deploy Pod + Service to K8s
    let ns = &state.namespace;
    let pod_name = format!("adapter-{}", req.name);
    let labels: BTreeMap<String, String> = BTreeMap::from([
        ("app".to_string(), pod_name.clone()),
        ("regret.io/adapter".to_string(), req.name.clone()),
        ("regret.io/adapter-id".to_string(), id.clone()),
    ]);

    let env_vars: Vec<EnvVar> = req.env.iter().map(|(k, v)| EnvVar {
        name: k.clone(),
        value: Some(v.clone()),
        ..Default::default()
    }).collect();

    let pod = Pod {
        metadata: ObjectMeta {
            name: Some(pod_name.clone()),
            namespace: Some(ns.clone()),
            labels: Some(labels.clone()),
            ..Default::default()
        },
        spec: Some(PodSpec {
            containers: vec![Container {
                name: "adapter".to_string(),
                image: Some(req.image.clone()),
                image_pull_policy: Some("Always".to_string()),
                ports: Some(vec![ContainerPort {
                    container_port: 9090,
                    ..Default::default()
                }]),
                env: Some(env_vars),
                ..Default::default()
            }],
            ..Default::default()
        }),
        ..Default::default()
    };

    let svc = Service {
        metadata: ObjectMeta {
            name: Some(pod_name.clone()),
            namespace: Some(ns.clone()),
            labels: Some(labels.clone()),
            ..Default::default()
        },
        spec: Some(ServiceSpec {
            selector: Some(BTreeMap::from([
                ("app".to_string(), pod_name.clone()),
            ])),
            ports: Some(vec![ServicePort {
                port: 9090,
                target_port: Some(IntOrString::Int(9090)),
                ..Default::default()
            }]),
            ..Default::default()
        }),
        ..Default::default()
    };

    let pods: Api<Pod> = Api::namespaced(state.kube.clone(), ns);
    let svcs: Api<Service> = Api::namespaced(state.kube.clone(), ns);
    let pp = PostParams::default();

    // Deploy pod
    if let Err(e) = pods.create(&pp, &pod).await {
        // Rollback DB record
        warn!(%e, "failed to create adapter pod, rolling back DB record");
        let _ = state.sqlite.delete_adapter(&id).await;
        return Err(ApiError::Internal(anyhow::anyhow!("failed to deploy adapter pod: {e}")));
    }

    // Deploy service
    if let Err(e) = svcs.create(&pp, &svc).await {
        // Rollback pod + DB record
        warn!(%e, "failed to create adapter service, rolling back");
        let _ = pods.delete(&pod_name, &DeleteParams::default()).await;
        let _ = state.sqlite.delete_adapter(&id).await;
        return Err(ApiError::Internal(anyhow::anyhow!("failed to deploy adapter service: {e}")));
    }

    info!(id = %id, name = %req.name, pod = %pod_name, "adapter deployed");
    Ok((StatusCode::CREATED, Json(to_response(&adapter))))
}

pub async fn list(
    State(state): State<AppState>,
) -> Result<impl IntoResponse, ApiError> {
    let adapters = state.sqlite.list_adapters_all().await?;
    let items: Vec<AdapterResponse> = adapters.iter().map(to_response).collect();
    Ok(Json(serde_json::json!({ "items": items })))
}

pub async fn get_one(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let adapter = state.sqlite.get_adapter(&id).await?
        .ok_or_else(|| ApiError::NotFound(format!("adapter {id} not found")))?;
    Ok(Json(to_response(&adapter)))
}

pub async fn delete(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let adapter = state.sqlite.get_adapter(&id).await?
        .ok_or_else(|| ApiError::NotFound(format!("adapter {id} not found")))?;

    let ns = &state.namespace;
    let pod_name = format!("adapter-{}", adapter.name);
    let dp = DeleteParams::default();

    // Delete K8s resources first (best-effort)
    let pods: Api<Pod> = Api::namespaced(state.kube.clone(), ns);
    let svcs: Api<Service> = Api::namespaced(state.kube.clone(), ns);

    if let Err(e) = svcs.delete(&pod_name, &dp).await {
        warn!(%e, "failed to delete adapter service (may not exist)");
    }
    if let Err(e) = pods.delete(&pod_name, &dp).await {
        warn!(%e, "failed to delete adapter pod (may not exist)");
    }

    // Delete DB record
    let deleted = state.sqlite.delete_adapter(&id).await?;
    if !deleted {
        return Err(ApiError::NotFound(format!("adapter {id} not found")));
    }

    info!(id = %id, name = %adapter.name, "adapter deleted");
    Ok(StatusCode::NO_CONTENT)
}

fn to_response(a: &crate::storage::sqlite::AdapterRecord) -> AdapterResponse {
    AdapterResponse {
        id: a.id.clone(),
        name: a.name.clone(),
        image: a.image.clone(),
        env: serde_json::from_str(&a.env).unwrap_or_default(),
        created_at: a.created_at.clone(),
    }
}

File: pilot/endpoints/adapters.go
package endpoints

import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"

	"github.com/go-chi/chi/v5"
	apierrors "k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/regret-io/regret/pilot-go/database"
	"github.com/regret-io/regret/pilot-go/ext"
)

type adapterHandlers struct {
	state *AppState
}

// Create handles POST /api/adapters.
func (a *adapterHandlers) Create(w http.ResponseWriter, r *http.Request) {
	var req CreateAdapterRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		WriteError(w, BadRequest("invalid JSON: "+err.Error()))
		return
	}

	ctx := r.Context()

	cfg, configYAML, err := adapterConfigFromRequest(&req)
	if err != nil {
		WriteError(w, BadRequest(err.Error()))
		return
	}
	envJSON, _ := json.Marshal(adapterEnvJSONMap(cfg))

	// Check name uniqueness
	existing, _ := a.state.Sqlite.GetAdapterByName(ctx, req.Name)
	if existing != nil {
		WriteError(w, Conflict(fmt.Sprintf("adapter '%s' already exists", req.Name)))
		return
	}

	id := fmt.Sprintf("adp-%d", ext.NowUnixMilli())

	// Step 1: Save to DB
	adapter, err := a.state.Sqlite.CreateAdapter(ctx, id, req.Name, cfg.Image, string(envJSON), configYAML)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}

	// Step 2: Deploy Pod + Service to K8s (if kube client is available)
	if a.state.KubeClient != nil {
		if err := a.deployAdapterToK8s(req.Name, id, cfg); err != nil {
			// Rollback DB record
			slog.Warn("failed to deploy adapter, rolling back", slog.Any("error", err))
			_, _ = a.state.Sqlite.DeleteAdapter(ctx, id)
			WriteError(w, InternalError(fmt.Errorf("failed to deploy adapter: %w", err)))
			return
		}
	}

	WriteJSON(w, http.StatusCreated, toAdapterResponse(adapter))
}

// List handles GET /api/adapters.
func (a *adapterHandlers) List(w http.ResponseWriter, r *http.Request) {
	adapters, err := a.state.Sqlite.ListAdapters(r.Context())
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}

	items := make([]AdapterResponse, 0, len(adapters))
	for i := range adapters {
		items = append(items, toAdapterResponse(&adapters[i]))
	}
	WriteJSON(w, http.StatusOK, map[string]interface{}{"items": items})
}

// GetOne handles GET /api/adapters/{id}.
func (a *adapterHandlers) GetOne(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	adapter, err := a.state.Sqlite.GetAdapter(r.Context(), id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("adapter %s not found", id)))
		return
	}
	WriteJSON(w, http.StatusOK, toAdapterResponse(adapter))
}

// Delete handles DELETE /api/adapters/{id}.
func (a *adapterHandlers) Delete(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	ctx := r.Context()

	adapterRec, err := a.state.Sqlite.GetAdapter(ctx, id)
	if err != nil {
		WriteError(w, NotFound(fmt.Sprintf("adapter %s not found", id)))
		return
	}

	// Delete K8s resources (best-effort)
	if a.state.KubeClient != nil {
		a.deleteAdapterFromK8s(adapterRec.Name)
	}

	deleted, err := a.state.Sqlite.DeleteAdapter(ctx, id)
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}
	if !deleted {
		WriteError(w, NotFound(fmt.Sprintf("adapter %s not found", id)))
		return
	}

	w.WriteHeader(http.StatusNoContent)
}

func (a *adapterHandlers) deployAdapterToK8s(adapterName, id string, cfg *AdapterConfig) error {
	service := buildAdapterService(a.state.Namespace, id, adapterName)
	deployment := buildAdapterDeployment(a.state.Namespace, id, adapterName, cfg)

	if _, err := a.state.KubeClient.CoreV1().Services(a.state.Namespace).Create(context.Background(), service, metav1.CreateOptions{}); err != nil && !apierrors.IsAlreadyExists(err) {
		return err
	}
	if _, err := a.state.KubeClient.AppsV1().Deployments(a.state.Namespace).Create(context.Background(), deployment, metav1.CreateOptions{}); err != nil {
		if !apierrors.IsAlreadyExists(err) {
			return err
		}
		if _, err := a.state.KubeClient.AppsV1().Deployments(a.state.Namespace).Update(context.Background(), deployment, metav1.UpdateOptions{}); err != nil {
			return err
		}
	}
	return nil
}

// deleteAdapterFromK8s removes the adapter Pod + Service.
func (a *adapterHandlers) deleteAdapterFromK8s(name string) {
	ctx := context.Background()
	_ = a.state.KubeClient.AppsV1().Deployments(a.state.Namespace).Delete(ctx, adapterWorkloadName(name), metav1.DeleteOptions{})
	_ = a.state.KubeClient.CoreV1().Services(a.state.Namespace).Delete(ctx, adapterWorkloadName(name), metav1.DeleteOptions{})
}

func toAdapterResponse(a *database.AdapterRecord) AdapterResponse {
	var env interface{}
	if err := json.Unmarshal([]byte(a.Env), &env); err != nil {
		env = map[string]interface{}{}
	}
	configYAML := a.ConfigYAML
	if configYAML == "" {
		envMap := map[string]string{}
		_ = json.Unmarshal([]byte(a.Env), &envMap)
		if generated, err := legacyAdapterConfigYAML(a.Image, envMap); err == nil {
			configYAML = generated
		}
	}
	return AdapterResponse{
		ID:         a.ID,
		Name:       a.Name,
		ConfigYAML: configYAML,
		Image:      a.Image,
		Env:        env,
		CreatedAt:  a.CreatedAt,
	}
}
